src/apps/lwaftr/address_map.lua

```diff
@@ -0,0 +1,313 @@
+-- Address maps
+--
+-- The lw4o6 architecture supports sharing of IPv4 addresses by
+-- partitioning the space of TCP/UDP/ICMP ports into disjoint "port
+-- sets".  Each softwire associated with an IPv4 address corresponds to
+-- a different set of ports on that address.  The way that the ports are
+-- partitioned is specified in RFC 7597: each address has an associated
+-- set of parameters that specifies how to compute a "port set
+-- identifier" (PSID) from a given port.
+--
+--                      0                   1
+--                      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
+--                     +-----------+-----------+-------+
+--       Ports in      |     A     |    PSID   |   j   |
+--    the CE port set  |    > 0    |           |       |
+--                     +-----------+-----------+-------+
+--                     |  a bits   |  k bits   |m bits |
+--
+--             Figure 2: Structure of a Port-Restricted Port Field
+--
+-- Source: http://tools.ietf.org/html/rfc7597#section-5.1 
+--
+-- We find the specification's names to be a bit obtuse, so we refer to
+-- them using the following names:
+--
+--   a bits = reserved_ports_bit_count.
+--   k bits = psid_length.
+--   m bits = shift.
+--
+-- Anyway, an address map is a lookup table that, given an IPv4 address
+-- and a port, uses the appropriate "psid_length" and "shift" parameters
+-- to compute a PSID.  If the IPv4 address is not under control of the
+-- lwAFTR, the address map still returns a PSID, under the assumption
+-- that the subsequent binding table lookup will fail.  After all, what
+-- we're really interested in is mapping a packet to a binding table
+-- entry, and computing the PSID is just a detail.
+-- 
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local rangemap = require("apps.lwaftr.rangemap")
+
+local address_map_value = ffi.typeof([[
+   struct { uint16_t psid_length; uint16_t shift; }
+]])
+
+local Parser = {}
+
+function Parser.new(file)
+   local name = '<unknown>'
+   if type(file) == 'string' then name, file = file, io.open(file) end
+   local ret = { column=0, line=0, name=name }
+   function ret.read_char() return file:read(1) end
```

Comment:
I think it's necessary a method to close the file descriptor. I think a good point from where to call such method would be error, as it finalizes input reading from the file descriptor.

```lua
function ret.close() return file:close() end
```

```lua
function Parser:error(msg, ...)
   self:close()
   error(...)
end
```